### PR TITLE
Banners

### PR DESCRIPTION
**Pull request by ZephireNZ**
API changes to for Banners in 1.8.

The major classes:
- BannerPattern: This is a combination of _all_ the patterns on the banner. Each pattern type has a related color. It returns a Map of Type to Color, where the iteration order is dependable (it iterates from the bottom layer tothe top, as expected.
- block.Banner: The BlockState for the banner, which has the base color (a dye color), and a pattern (as above).
- material.Banner: The _blocks_ MaterialData, which is the rotation of the block. This should really be merged with Sign, as they use the same values, but I can't be assed :P
- meta.BannerMeta: The ItemMeta for the item. This is the values stored in the BlockEntityTag - see GlowstoneMC/Glowstone#326 for how this could be generified for all blocks.

### CODE REVIEW

**Comment by Postremus** [comment]
maybe the getFacing and setFacing methods should use the same code as signs does

File: src/main/java/org/bukkit/BannerPattern.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+        private Type(String code) {
+            this.code = code;
+        }
+
+        /**
+         * Gets the internal code of this banner pattern
+         * @return The 2/3 character pattern code
+         */
+        public String getCode() {
```
Comment (Tonodus) [inline src/main/java/org/bukkit/BannerPattern.java:71]: Should be depraced, since this is a magic value. Same goes for #getByCode()

**Comment by ZephireNZ** [comment]
Yeah. I looked at the rest of the Bukkit code, however, and all the objects that implement Directional also have their own separate implementations of getFacing and setFacingDirection, despite being similar. Not saying it's _good_ practice, but I was just following what the rest of the API was doing.

File: src/main/java/org/bukkit/BannerPattern.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        CROSS("cr"),
+        CURLY_BORDER("cbo"),
+        DIAGONAL_LEFT("ld"),
+        DIAGONAL_RIGHT("rud"),
+        DIAGONAL_UP_LEFT("lud"),
+        DIAGONAL_UP_RIGHT("rd"),
+        FLOWER("flo"),
+        GRADIENT("gra"),
+        GRADIENT_UP("gru"),
+        HALF_HORIZONTAL("hh"),
```
Comment (turt2live) [inline src/main/java/org/bukkit/BannerPattern.java]: Name suggestion: `TOP_HALF`

File: src/main/java/org/bukkit/BannerPattern.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        HALF_HORIZONTAL_BOTTOM("hhb"),
```
Comment (turt2live) [inline src/main/java/org/bukkit/BannerPattern.java]: Name suggestion: `BOTTOM_HALF`

File: src/main/java/org/bukkit/BannerPattern.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        LEFT_HALF("vh"),
+        LEFT_HALF_RIGHT("vhr"),
```
Comment (turt2live) [inline src/main/java/org/bukkit/BannerPattern.java]: Name suggestion: `RIGHT_HALF`

File: src/main/java/org/bukkit/BannerPattern.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        MOJANG("moj"),
+        RHOMBUS("mr"),
+        SKULL("sku"),
+        SMALL_STRIPES("ss"),
+        SQUARE_BOTTOM_LEFT("bl"),
+        SQUARE_BOTTOM_RIGHT("br"),
+        SQUARE_TOP_LEFT("tl"),
+        SQUARE_TOP_RIGHT("tr"),
+        STRAIGHT_CROSS("sr"),
```
Comment (turt2live) [inline src/main/java/org/bukkit/BannerPattern.java]: Does not function

File: src/main/java/org/bukkit/inventory/meta/BannerMeta.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * Represents a banner item.
+ */
+public interface BannerMeta extends ItemMeta {
+
+    /**
+     * Sets the pattern of the banner. The old pattern is removed completely.
+     * @param pattern Pattern to change to
+     * @throws IllegalArgumentException If pattern is null
+     */
+    public void setPattern(BannerPattern pattern) throws NullPointerException;
```
Comment (turt2live) [inline src/main/java/org/bukkit/inventory/meta/BannerMeta.java]: throws is unrequired

**Comment by ZephireNZ** [comment]
Pushing this back to WIP, turns out I didn't quite thinks things through, specifically to do with using a Map for something that can have duplicate layers.

**Comment by ZephireNZ** [comment]
Fixed :3

**Comment by SpaceManiac** [comment]
Merged in 162c1c3.

**Pull request closed without merging**